src/annoylib.h

```diff
@@ -473,28 +475,78 @@ template<typename S, typename T, typename Distance, typename Random>
       _n_items = item + 1;
   }
 
-  void build(int q) {
+  void build_tree_threaded(int num_trees, vector<S> &_roots) {
+	  // just for this thread
+	  vector<S> _roots_thread;
+	  while (1) {
+		  // user doesn't specify number of trees,
+		  // just fully develop one tree?
+		  if (num_trees == -1 && _n_nodes >= _n_items * 2)
+			  break;
+		  // quit when we have all trees built for this thread
+		  if (num_trees != -1 && _roots_thread.size() >= (size_t)num_trees)
+			  break;
+		  if (_verbose) showUpdate("pass %zd...\n", _roots.size());
+
+		  vector<S> indices;
+		  for (S i = 0; i < _n_items; i++) {
+			  if (_get(i)->n_descendants >= 1) // Issue #223
+				  indices.push_back(i);
+		  }
+		  S tree = _make_tree(indices);
+		  _roots_thread.push_back(tree);
+	  }
+	  // push to our shared roots vector
+	  _roots.insert(_roots.end(), _roots_thread.begin(), _roots_thread.end());
```

Comment:
this needs to be protected by a mutex.

Reply:
It shouldn't be using boost python (I think an earlier version of annoy did though).

Reply:
Oh, that's right... I forgot. Hm.

Reply:
seems easier if each thread just returns a vector of roots and the main thread pushes them into the roots array